runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1281 @@
+		#[pallet::constant]
+		type MaxTemporarySlots: Get<u32>;
+
+		/// The max number of temporary slots to be scheduled per lease periods.
+		#[pallet::constant]
+		type MaxTemporarySlotPerLeasePeriod: Get<u32>;
+	}
+
+	/// Assigned permanent slots, with their start lease period.
+	#[pallet::storage]
+	#[pallet::getter(fn permanent_slots)]
+	pub type PermanentSlots<T: Config> =
+		StorageMap<_, Twox64Concat, ParaId, Option<LeasePeriodOf<T>>, ValueQuery>;
+
+	/// Number of assigned (and active) permanent slots.
+	#[pallet::storage]
+	#[pallet::getter(fn permanent_slot_count)]
+	pub type PermanentSlotCount<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// Assigned temporary slots.
+	#[pallet::storage]
+	#[pallet::getter(fn temporary_slots)]
+	pub type TemporarySlots<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		ParaId,
+		Option<ParachainTemporarySlot<T::AccountId, LeasePeriodOf<T>>>,
+		ValueQuery,
+	>;
+
+	/// Number of assigned temporary slots.
+	#[pallet::storage]
+	#[pallet::getter(fn temporary_slot_count)]
+	pub type TemporarySlotCount<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// Number of active temporary slots in current slot lease period.
+	#[pallet::storage]
+	#[pallet::getter(fn active_temporary_slot_count)]
+	pub type ActiveTemporarySlotCount<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A para was assigned a permanent parachain slot
+		PermanentSlotAssigned(ParaId),
+		/// A para was assigned a temporary parachain slot
+		TemporarySlotAssigned(ParaId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The specified parachain or parathread is not registered.
+		ParaDoesntExist,
+		/// Not a parathread.
+		NotParathread,
+		/// Cannot upgrade parathread.
+		CannotUpgrade,
+		/// Cannot downgrade parachain.
+		CannotDowngrade,
+		/// Permanent or Temporary slot already assigned.
+		SlotAlreadyAssigned,
+		/// Permanent or Temporary slot has not been assigned.
+		SlotNotAssigned,
+		/// An ongoing lease already exists.
+		OngoingLeaseExists,
+		// Maximum number of permanent slots exceeded
+		MaxPermanentSlotsExceeded,
+		// Maximum number of temporary slots exceeded
+		MaxTemporarySlotsExceeded,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			if let Some((lease_period, first_block)) = Self::lease_period_index(n) {
+				// If we're beginning a new lease period then handle that.
+				if first_block {
+					return Self::manage_lease_period_start(lease_period)
+				}
+			}
+
+			// We didn't return early above, so we didn't do anything.
+			0
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Assign a permanent parachain slot and immediately create a lease for it.
+		#[pallet::weight((1_000 + <T as slots::Config>::WeightInfo::force_lease(), DispatchClass::Operational))]
```

Comment:
If you don't have real benchmarks, better to set this to something like 1/10th the total block weight, which is certainly an over-estimate.

Reply:
and add a todo